src/main/scala/MilvusDataReader.scala

```diff
@@ -15,18 +15,84 @@ object MilvusDataReader {
       "s3a://"
     )
 
-    val insertDF = spark.read
-      .format("milvus")
-      .options(config.options)
-      .option(
-        MilvusOption.S3FileSystemTypeName,
-        fileSystemType
+    // Detect storage versions by querying segment info
+    val client = MilvusClient(
+      MilvusConnectionParams(
+        uri = config.uri,
+        token = config.token,
+        databaseName = config.options.getOrElse(MilvusOption.MilvusDatabaseName, "")
       )
-      .option(MilvusOption.ReaderType, "insert")
-      .option(MilvusOption.MilvusUri, config.uri)
-      .option(MilvusOption.MilvusToken, config.token)
-      .option(MilvusOption.MilvusCollectionName, config.collectionName)
-      .load()
+    )
+
+    val (hasV1, hasV2) = try {
+      val segments = client.getSegments(
+        config.options.getOrElse(MilvusOption.MilvusDatabaseName, ""),
+        config.collectionName
+      ).getOrElse(Seq.empty)
+
+      val v1Count = segments.count(_.storageVersion < 2)
+      val v2Count = segments.count(_.storageVersion >= 2)
+
+      println(s"Collection ${config.collectionName}: Total segments=${segments.size}, V1=${v1Count}, V2=${v2Count}")
+
+      (v1Count > 0, v2Count > 0)
+    } catch {
+      case e: Exception =>
+        println(s"Warning: Failed to detect storage version, defaulting to v1: ${e.getMessage}")
+        (true, false)
+    } finally {
+      client.close()
+    }
+
+    // Read from appropriate format(s) based on segment versions
+    val insertDF = if (hasV1 && hasV2) {
+      // Mixed version collection - read from both formats and union
+      println(s"Mixed version collection detected, reading from both milvus and storagev2 formats")
+
+      val v1DF = spark.read
+        .format("milvus")
+        .options(config.options)
+        .option(MilvusOption.S3FileSystemTypeName, fileSystemType)
+        .option(MilvusOption.ReaderType, "insert")
+        .option(MilvusOption.MilvusUri, config.uri)
+        .option(MilvusOption.MilvusToken, config.token)
+        .option(MilvusOption.MilvusCollectionName, config.collectionName)
+        .load()
+
+      val v2DF = spark.read
+        .format("storagev2")
+        .options(config.options)
+        .option(MilvusOption.S3FileSystemTypeName, fileSystemType)
+        .option(MilvusOption.MilvusUri, config.uri)
+        .option(MilvusOption.MilvusToken, config.token)
+        .option(MilvusOption.MilvusCollectionName, config.collectionName)
+        .load()
+
+      v1DF.union(v2DF)
+    } else if (hasV2) {
+      // Only V2 segments
+      println(s"V2 collection detected, using storagev2 format")
+      spark.read
+        .format("storagev2")
```

Comment:
We should not make this a different format. We only have one format registered, that is `milvus`